pilot/pkg/networking/core/v1alpha3/cluster.go

```diff
@@ -515,6 +515,10 @@ var h2UpgradeMap = map[upgradeTuple]bool{
 func applyH2Upgrade(opts buildClusterOpts, connectionPool *networking.ConnectionPoolSettings) {
 	if shouldH2Upgrade(opts.cluster.Name, opts.direction, opts.port, opts.mesh, connectionPool) {
 		setH2Options(opts.cluster)
+		// Use downstream protocol. If the incoming traffic use HTTP 1.1, the
+		// upstream cluster will use HTTP 1.1, if incoming traffic use HTTP2,
```

Comment:
This actually defeats upgrade.

Client (h1) — client sidecar (upgrade to h2) 
Is what we want.
Use-downstream here will cause the upgrade to not happen.

Reply:
Correct. This will prevent the upgrade to happen. But what how do we express to preserve the client protocol? May be not part of the upgrade option -  some other different config? 

BTW, we do similar thing for sidecar already https://github.com/istio/istio/blob/master/pilot/pkg/networking/core/v1alpha3/cluster.go#L1044.

We are looking to get the same behaviour for gateways. It was originally added for gateways as well - but was reverted for https://github.com/istio/istio/issues/21230 this issue.